src/compiler/checker.ts

```diff
@@ -44536,31 +44536,39 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         checkSourceElement(node.elseStatement);
     }
 
-    function checkTestingKnownTruthyCallableOrAwaitableOrEnumMemberType(condExpr: Expression, condType: Type, body?: Statement | Expression) {
+    function checkTestingKnownTruthyCallableOrAwaitableOrEnumMemberType(inputCondExpr: Expression, condType: Type, body?: Statement | Expression) {
         if (!strictNullChecks) return;
-        bothHelper(condExpr, body);
 
-        function bothHelper(condExpr: Expression, body: Expression | Statement | undefined) {
-            condExpr = skipParentheses(condExpr);
-
-            helper(condExpr, body);
+        bothHelper(inputCondExpr);
 
+        function bothHelper(condExpr: Expression) {
+            condExpr = skipParentheses(condExpr);
+            helper(condExpr);
             while (isBinaryExpression(condExpr) && (condExpr.operatorToken.kind === SyntaxKind.BarBarToken || condExpr.operatorToken.kind === SyntaxKind.QuestionQuestionToken)) {
                 condExpr = skipParentheses(condExpr.left);
-                helper(condExpr, body);
+                helper(condExpr);
             }
         }
 
-        function helper(condExpr: Expression, body: Expression | Statement | undefined) {
-            const location = isLogicalOrCoalescingBinaryExpression(condExpr) ? skipParentheses(condExpr.right) : condExpr;
+        function helper(condExpr: Expression) {
+            let inverted = false;
+            let location = condExpr;
+            if (isPrefixUnaryExpression(condExpr) && condExpr.operator === SyntaxKind.ExclamationToken) {
+                if (!isIfStatement(condExpr.parent)) {
+                    return;
+                }
+                location = skipParentheses(condExpr.operand);
+                inverted = true;
+            }
+            location = isLogicalOrCoalescingBinaryExpression(location) ? skipParentheses(location.right) : location;
             if (isModuleExportsAccessExpression(location)) {
                 return;
             }
             if (isLogicalOrCoalescingBinaryExpression(location)) {
-                bothHelper(location, body);
+                bothHelper(location);
                 return;
             }
-            const type = location === condExpr ? condType : checkExpression(location);
+            const type = location === inputCondExpr ? condType : checkExpression(location);

```

Comment:
I think this was a bug before? It's an optimisation to not need to call `checkExpression` to get the `type` and use the provided `condType`, if the location still matched the provided `condExpr`?